### ISSUE TITLE
False positive `used-before-assignment` for type annotations that take assignment via nonlocal

### ISSUE THREAD

**Issue by IRDonch** [PyCQA/pylint#5394]
### Bug description

Create `test.py` with the following content:

```python
def f():
    x: int

    def g():
        nonlocal x
        x = 5

    g()
    print(x)

f()
```

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:9:10: E0602: Undefined variable 'x' (undefined-variable)

------------------------------------------------------------------
Your code has been rated at 3.75/10 (previous run: 3.75/10, +0.00)
```

### Expected behavior

There should be no errors reported.

### Pylint version

```shell
pylint 2.12.1
astroid 2.9.0
Python 3.6.9 (default, Jan 26 2021, 15:33:00)
[GCC 8.4.0]
```

**Comment by IRDonch** [comment]
This is a regression in pylint 2.12. Probably caused by #5158.

**Comment by DanielNoord** [comment]
This might be difficult to solve as it requires control-flow knowledge. We can't do a simple check for a definition in an inner function as then this wouldn't be catched:

```python
def f():
    x: int

    def g():
        nonlocal x
        x = 5

    print(x) # x is undefined
    g()

f()
```

For now I would advise to add an disable to this line. Sorry about that!

**Comment by IRDonch** [comment]
I understand that it's undecidable in the general case. You could, however, err on the side of avoiding false positives by only emitting this error if the variable is not assigned in any scope.

**Comment by DanielNoord** [comment]
I don't remember the code by heart but I'll have a look in the near future. I'm not sure whether that approach would greatly increase the amount of false negatives, but I/we can at least try.

If somebody else would like to have a look before me, please go ahead!

**Comment by jacobtylerwalls** [comment]
On 2.13.x this emits `used-before-assignment` instead due to #5718
